controllers/gcpcluster_controller.go

```diff
@@ -45,51 +46,51 @@ type GCPClusterReconciler struct {
 	client.Client
 	Log              logr.Logger
 	ReconcileTimeout time.Duration
+	WatchFilterValue string
 }
 
-func (r *GCPClusterReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+func (r *GCPClusterReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
 	c, err := ctrl.NewControllerManagedBy(mgr).
 		WithOptions(options).
 		For(&infrav1.GCPCluster{}).
 		Watches(
 			&source.Kind{Type: &infrav1.GCPMachine{}},
-			&handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(r.GCPMachineToGCPCluster)},
+			handler.EnqueueRequestsFromMapFunc(r.GCPMachineToGCPCluster),
 		).
-		WithEventFilter(pausePredicates).Build(r)
+		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue)).
+		Build(r)
 	if err != nil {
 		return errors.Wrap(err, "error creating controller")
 	}
 
 	return c.Watch(
 		&source.Kind{Type: &clusterv1.Cluster{}},
-		&handler.EnqueueRequestsFromMapFunc{
-			ToRequests: handler.ToRequestsFunc(r.requeueGCPClusterForUnpausedCluster),
-		},
-		predicate.Funcs{
-			CreateFunc: func(e event.CreateEvent) bool {
-				cluster := e.Object.(*clusterv1.Cluster)
-
-				return !cluster.Spec.Paused
-			},
-			DeleteFunc: func(e event.DeleteEvent) bool {
-				return false
-			},
-			UpdateFunc: func(e event.UpdateEvent) bool {
-				oldCluster := e.ObjectOld.(*clusterv1.Cluster)
-				newCluster := e.ObjectNew.(*clusterv1.Cluster)
-
-				return oldCluster.Spec.Paused && !newCluster.Spec.Paused
-			},
-		},
+		handler.EnqueueRequestsFromMapFunc(util.ClusterToInfrastructureMapFunc(infrav1.GroupVersion.WithKind("GCPCluster"))),
+		// predicate.Funcs{
```

Comment:
Dont know how to deal with this part

Reply:
This should be able to be consolidated into the block above, since the newer version of controller-runtime you can specify predicates on watches in the builder instead of having to do it afterwards...

I think it can look something like this:

```go
func (r *GCPClusterReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
	return ctrl.NewControllerManagedBy(mgr).
		WithOptions(options).
		For(&infrav1.GCPCluster{}).
		Watches(
			&source.Kind{Type: &infrav1.GCPMachine{}},
			handler.EnqueueRequestsFromMapFunc(r.GCPMachineToGCPCluster),
		).
		Watches(
			&source.Kind{Type: &clusterv1.Cluster{}},
			handler.EnqueueRequestsFromMapFunc(util.ClusterToInfrastructureMapFunc(infrav1.GroupVersion.WithKind("GCPCluster"))),
			WithPredicates(predicates.ClusterUnpaused(ctrl.LoggerFrom(ctx))),
		).
		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue)).
		Complete(r)
```